lib/ruby_lsp/requests/document_highlight.rb

```diff
@@ -0,0 +1,153 @@
+            else
+              position <=> child.location.start_char
+            end
+          end
+
+        case matched
+        when SyntaxTree::GVar
+          GVarHighlightVisitor.new(matched)
+        when SyntaxTree::Ident
+          IdentHighlightVisitor.new(matched)
+        when SyntaxTree::IVar
+          IVarHighlightVisitor.new(matched)
+        when SyntaxTree::Const
+          ConstHighlightVisitor.new(matched)
+        when SyntaxTree::Node
+          find(matched, position)
+        end
+      end
+
+      class HighlightVisitor < SyntaxTree::Visitor
+        attr_reader :highlights
+
+        def initialize(target)
+          @highlights = []
+          @assign_node_stack = []
+          @target = target
+
+          super()
+        end
+
+        def visit_assign(node)
+          @assign_node_stack << node
+          super
+        ensure
+          @assign_node_stack.pop
+        end
+
+        def visit_var_field(node)
+          process_node(node.value) if matches_target(node.value)
+        end
+
+        def visit_var_ref(node)
+          process_node(node.value) if matches_target(node.value)
+        end
+
+        private
+
+        WRITE_KIND = 3
+        READ_KIND = 2
+
+        def process_node(node)
+          if node.value == @target.value
+            kind = in_assignment?(node) ? WRITE_KIND : READ_KIND
+            highlight(node, kind)
+          end
+        end
+
+        def in_assignment?(node)
+          (assign_node = @assign_node_stack.last) && node == assign_node.target.value
+        end
+
+        def highlight(node, kind)
+          range = range_from_syntax_tree_node(node)
+          highlight = LanguageServer::Protocol::Interface::DocumentHighlight.new(range: range, kind: kind)
+          @highlights << highlight
+        end
+
+        def range_from_syntax_tree_node(node)
+          loc = node.location
+
+          LanguageServer::Protocol::Interface::Range.new(
+            start: LanguageServer::Protocol::Interface::Position.new(line: loc.start_line - 1,
+              character: loc.start_column),
+            end: LanguageServer::Protocol::Interface::Position.new(line: loc.end_line - 1, character: loc.end_column),
+          )
+        end
+      end
+
+      class IVarHighlightVisitor < HighlightVisitor
+        def matches_target(node)
+          node.is_a?(SyntaxTree::IVar) && node.value == @target.value
+        end
+      end
+
+      class GVarHighlightVisitor < HighlightVisitor
+        def matches_target(node)
+          node.is_a?(SyntaxTree::GVar) && node.value == @target.value
+        end
+      end
+
+      class IdentHighlightVisitor < HighlightVisitor
+        def matches_target(node)
+          node.is_a?(SyntaxTree::Ident) && node.value == @target.value
+        end
+      end
+
+      class ConstHighlightVisitor < HighlightVisitor
+        def matches_target(node)
+          node.is_a?(SyntaxTree::Const) && node.value == @target.value
+        end
+      end
```

Comment:
Right now the each sub-visitor and their `matches_target` methods look duplicated. But in the follow-up PRs each of them will have different scope-resolving implementations.